ecs-agent/acs/session/attach_instance_eni_responder.go

```diff
@@ -0,0 +1,154 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package session
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/pkg/errors"
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/acs/model/ecsacs"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/api/attachmentinfo"
+	apieni "github.com/aws/amazon-ecs-agent/ecs-agent/api/eni"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/api/status"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger/field"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/wsclient"
+)
+
+const (
+	AttachInstanceENIMessageName = "AttachInstanceNetworkInterfacesMessage"
+)
+
+// attachInstanceENIResponder implements the wsclient.RequestResponder interface for responding
+// to ecsacs.AttachInstanceNetworkInterfacesMessage messages sent by ACS.
+type attachInstanceENIResponder struct {
+	eniHandler ENIHandler
+	respond    wsclient.RespondFunc
+}
+
+// NewAttachInstanceENIResponder returns an instance of the attachInstanceENIResponder struct.
+func NewAttachInstanceENIResponder(eniHandler ENIHandler,
+	responseSender wsclient.RespondFunc) wsclient.RequestResponder {
+	r := &attachInstanceENIResponder{
+		eniHandler: eniHandler,
+	}
+	r.respond = ResponseToACSSender(r.Name(), responseSender)
+	return r
+}
+
+func (*attachInstanceENIResponder) Name() string { return "attach instance ENI responder" }
+
+func (r *attachInstanceENIResponder) HandlerFunc() wsclient.RequestHandler {
+	return r.handleAttachMessage
+}
+
+func (r *attachInstanceENIResponder) handleAttachMessage(message *ecsacs.AttachInstanceNetworkInterfacesMessage) {
+	logger.Debug(fmt.Sprintf("Handling %s", AttachInstanceENIMessageName))
+	receivedAt := time.Now()
+
+	// Validate fields in the message.
+	if err := validateAttachInstanceNetworkInterfacesMessage(message); err != nil {
+		logger.Error(fmt.Sprintf("Error validating %s received from ECS", AttachInstanceENIMessageName), logger.Fields{
+			field.Error: err,
+		})
+		return
+	}
+
+	// Handle ENIs in the message.
+	messageID := aws.StringValue(message.MessageId)
+	for _, mENI := range message.ElasticNetworkInterfaces {
+		expiresAt := receivedAt.Add(time.Duration(aws.Int64Value(message.WaitTimeoutMs)) * time.Millisecond)
+		go func(eni *ecsacs.ElasticNetworkInterface) {
```

Comment:
nit: any reason we need to create this inline anonymous function instead of having a separate function to do handling? We can then test that function as needed.

Reply:
That's a good point - I don't see any reason in particular that we can't have a separate function to do the handling. I'll make an update to the PR to do this instead.

Reply:
Sounds good.